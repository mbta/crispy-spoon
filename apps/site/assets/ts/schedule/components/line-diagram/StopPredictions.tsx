import React from "react";
import { capitalize } from "lodash";
import { HeadsignWithCrowding } from "../../../__v3api";
import {
  timeForCommuterRail,
  statusForCommuterRail
} from "../../../helpers/prediction-helpers";
import { isSkippedOrCancelled } from "../../../models/prediction";
import LiveCrowdingIcon from "./LiveCrowdingIcon";

interface StopPredictions {
  headsigns: HeadsignWithCrowding[];
  isCommuterRail: boolean;
}

export const hasPredictionTime = ({
  time_data_with_crowding_list: timeDataList
}: HeadsignWithCrowding): boolean =>
  !!(
    timeDataList &&
    timeDataList[0] &&
    timeDataList[0].time_data &&
    timeDataList[0].time_data.prediction &&
    timeDataList[0].time_data.prediction.time
  );

const predictionTexts = (headsign: HeadsignWithCrowding): string[] => {
  const texts = [headsign.name];
  if (headsign.train_number) {
    texts.push(` · Train ${headsign.train_number}`);
  }

  const predictedOrScheduledTime =
    headsign.time_data_with_crowding_list[0].time_data;
  const { prediction } = predictedOrScheduledTime;

  if (prediction && prediction.track) {
    texts.push(` · Track ${prediction.track}`);
  }

  const status = statusForCommuterRail(predictedOrScheduledTime);
  if (status) {
    texts.push(` · ${status}`);
  }
  return texts;
};

const StopPredictions = ({
  headsigns,
  isCommuterRail
}: StopPredictions): JSX.Element => {
  let predictions: JSX.Element[];
  const liveHeadsigns = headsigns.filter(hasPredictionTime);

  if (isCommuterRail) {
    // Display at most 1 prediction for Commuter Rail
    predictions = liveHeadsigns.slice(0, 1).map(headsign => {
      const enhancedTimeData = headsign.time_data_with_crowding_list[0];
      const predictedOrScheduledTime = enhancedTimeData.time_data;
      const predictionTimeClass = isSkippedOrCancelled(
        enhancedTimeData.predicted_schedule.prediction
      )
        ? "m-schedule-diagram__cr-prediction-time strikethrough"
        : "m-schedule-diagram__cr-prediction-time";

      return (
        <div key={`${headsign.name}-cr`}>
          <div className="m-schedule-diagram__cr-prediction">
            {timeForCommuterRail(predictedOrScheduledTime, predictionTimeClass)}
          </div>
          <div className="m-schedule-diagram__cr-prediction-details">
            {predictionTexts(headsign).map((text: string) => (
              <span key={text}>{text}</span>
            ))}
          </div>
        </div>
      );
    });
  } else {
    predictions = liveHeadsigns.map(headsign => {
      const { crowding } = headsign.time_data_with_crowding_list[0];
      return (
        /* eslint-disable-next-line react/no-array-index-key */
        <div key={headsign.name} className="m-schedule-diagram__prediction">
          <div>{headsign.name}</div>
          <div className="m-schedule-diagram__prediction-time">
            {capitalize(
              headsign.time_data_with_crowding_list[0].time_data.prediction!.time.join(
                " "
              )
            )}
          </div>
          <LiveCrowdingIcon crowding={crowding} />
        </div>
      );
    });
  }

  return <div className="m-schedule-diagram__predictions">{predictions}</div>;
};

export default StopPredictions;
