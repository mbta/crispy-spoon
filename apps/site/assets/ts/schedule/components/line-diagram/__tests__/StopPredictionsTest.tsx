import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { HeadsignWithCrowding, Schedule } from "../../../../__v3api";
import StopPredictions from "../StopPredictions";
import { TripPrediction as Prediction } from "../../__trips";

const mockPrediction = {} as Prediction & { headsign: string };
const mockSchedule = {} as Schedule & { headsign: string };

describe("StopPredictions", () => {
  it("renders bus predictions", () => {
    const headsigns: HeadsignWithCrowding[] = [
      {
        name: "Harvard",
        time_data_with_crowding_list: [
          {
            crowding: null,
            predicted_schedule: {
              prediction: mockPrediction,
              schedule: mockSchedule
            },
            time_data: {
              delay: 0,
              prediction: {
                schedule_relationship: null,
                status: null,
                time: ["6", " ", "min"],
                track: null
              },
              scheduled_time: ["3:15", " ", "PM"]
            }
          }
        ],
        train_number: ""
      }
    ];

    const tree = renderer
      .create(<StopPredictions headsigns={headsigns} isCommuterRail={false} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders commuter rail predictions", () => {
    const headsigns: HeadsignWithCrowding[] = [
      {
        name: "Worcester",
        time_data_with_crowding_list: [
          {
            crowding: null,
            predicted_schedule: {
              prediction: mockPrediction,
              schedule: mockSchedule
            },
            time_data: {
              delay: 0,
              prediction: {
                schedule_relationship: null,
                status: null,
                time: ["6", " ", "min"],
                track: null
              },
              scheduled_time: ["4:15", " ", "PM"]
            }
          }
        ],
        train_number: "7519"
      }
    ];

    const tree = renderer
      .create(<StopPredictions headsigns={headsigns} isCommuterRail={true} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders commuter rail predictions with a skipped stop", () => {
    const skippedPrediction = {
      schedule_relationship: "skipped",
      time: "1:23 pm",
      headsign: ""
    } as Prediction & { headsign: string };
    const headsigns: HeadsignWithCrowding[] = [
      {
        name: "Worcester",
        time_data_with_crowding_list: [
          {
            crowding: null,
            predicted_schedule: {
              prediction: skippedPrediction,
              schedule: mockSchedule
            },
            time_data: {
              delay: 0,
              prediction: {
                schedule_relationship: null,
                status: null,
                time: ["6", " ", "min"],
                track: null
              },
              scheduled_time: ["4:15", " ", "PM"]
            }
          }
        ],
        train_number: "7519"
      }
    ];

    const wrapper = mount(
      <StopPredictions headsigns={headsigns} isCommuterRail={true} />
    );

    expect(
      wrapper
        .find(".m-schedule-diagram__cr-prediction-time.strikethrough")
        .exists()
    ).toBeTruthy();
  });
});
