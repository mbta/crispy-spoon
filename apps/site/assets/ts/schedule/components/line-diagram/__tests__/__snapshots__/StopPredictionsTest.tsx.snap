// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopPredictions renders bus predictions 1`] = `
<div
  className="m-schedule-diagram__predictions"
>
  <div
    className="m-schedule-diagram__prediction notranslate"
  >
    <div>
      Harvard
    </div>
    <div
      className="m-schedule-diagram__prediction-time"
    >
      6   min
    </div>
    <div
      className="m-schedule-diagram__prediction-crowding"
    >
      <span
        aria-hidden="true"
        className="notranslate c-svg__icon c-icon__crowding c-icon__crowding--crowding_unavailable"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`StopPredictions renders commuter rail predictions 1`] = `
<div
  className="m-schedule-diagram__predictions"
>
  <div>
    <div
      className="m-schedule-diagram__cr-prediction"
    >
      <div
        className="m-schedule-diagram__cr-prediction-time"
      >
        6 min
      </div>
    </div>
    <div
      className="m-schedule-diagram__cr-prediction-details"
    >
      <span>
        Worcester
      </span>
      <span>
         · Train 7519
      </span>
      <span>
         · On time
      </span>
    </div>
  </div>
</div>
`;
